README.md
# natto-py
A Tasty Python Binding with MeCab

## What is natto-py?
natto combines the [Python programming language](https://www.python.org/) with [MeCab](http://mecab.googlecode.com/svn/trunk/mecab/doc/index.html), the part-of-speech and morphological analyzer for the Japanese language.

You can learn more about [natto-py at bitbucket](https://bitbucket.org/buruzaemon/natto-py).

## Requirements
natto-py requires the following:

-  [MeCab _0.996_](http://code.google.com/p/mecab/downloads/list)
-  [cffi _0.8.6 or greater_](https://bitbucket.org/cffi/cffi)
-  Python _2.7_

## Changelog
Please see the {file:CHANGELOG} for the release history.

## Copyright
Copyright &copy; 2014, Brooke M. Fujita. All rights reserved. Please see the {file:LICENSE} file for further details. 


CHANGELOG
CHANGELOG
---------

- 2022/09/07 1.0.1 release.
    - Issue 122: Improve logging set-up
    - As Python 3.5 and 3.6 has reached EOL, removing their refs from documentation
    - Minor update for pytz version in doc requirements.txt
    - Now using github actions for automated tests 

- 2021/10/03 1.0.0 release.
    - Issue 119: Failed to install natto-py when setuptools 58.0 or later is used
    - Support for Python 2 has ended; please use 0.9.2 if you are using Python 2.

- 2020/01/01 0.9.2 release.
    - Issue 114: Unexpected whitespace inserted in Wakati mode 

- 2019/12/27 0.9.1 release.
    - Issue 112: Erroneous handling of whitespace trimming
    - Issue 113: Security vulnerability for Jinja (requirement for documentation generation) 
    - Adding Python 3.8 to Travis CI tests
    - Adding doc references to Python 3.8 in supported versions 

- 2018/01/30 0.9.0 release.
    - Issue 92: Where's the natto-py documentation?
    - Issue 96: Index out of range error message not being captured by natto-py
    - Issue 99: Node-formatting ignored when using Unidic unless -O is set to empty string
    - Issue 101: Enable -r/--rcfile MeCab option

- 2016/01/31 0.8.0 release.
    - Issue 72: Revamp IPython notebooks for use with Jupyter
    - Issue 73: Better examples for 制約付き解析
    - Issue 78: Dynamically set version and version info for docs as well
    - Issue 83: natto.api.MeCabError: MECAB_NBEST request type is not set
    - Issue 84: Start leveraging the logging library

- 2015/11/07 0.7.0 release.
    - Issue 67: Confirm behavior with Python 3.5
    - Issue 68: Update README with dependencies for dev & testing
    - Issue 69: Read in version string from file

- 2015/05/02 0.6.0 release.
    - Issue 34: Add support for feature constraint parsing 

- 2015/04/18 0.5.1 release.
    - Issue 48: Object destructor for MeCab dumps core in 0.5.0

- 2015/04/17 0.5.0 release.
    - Issue 35: Use new Model- and Lattice-based C APIs internally
    - Issue 36: Make refs to MeCab and Tagger consistent in docs
    - Issue 37: Partial parsing and handling of tokens
    - Issue 38: Downloads and license badges for README.md & API docs

- 2015/03/31 0.4.0 release.
    - Issue 21: Cleaned up Lattice-related API call.
    - Issue 23: Implemented support for theta option on Lattice.
    - Issue 25: Enhanced tokenization to remove whitespace for partial parse.
    - Issue 26: Updated to docs/index.rst for new MeCab home at github. 

- 2015/03/19 0.3.0 release.
    - Added morpheme boundary constraint parsing.
    - Minor corrections to docs/index.rst formatting. 

- 2015/03/19 0.3.0 release.
    - Pulled.

- 2015/02/22 0.2.1 release.
    - Enhancement for Issue 11: Debugging should use a kwarg. 
    - On-going improvements and clarifications to natto-py documentation.

- 2015/02/22 0.2.0 release.
    - Pulled.

- 2015/01/02: 0.1.0 release.
    - Issue 9: Refactored Python 23 support logic. 
    - On-going improvements and clarifications to natto-py documentation.

- 2014/12/31: 0.0.9 release.
    - MeCab: libpath attribute returns absolute path to MeCab library.
    - DictionaryInfo: filepath attribute returns absolute path to dictionary.
    - Improvements and clarifications to natto-py documentation.

- 2014/12/02: 0.0.8 release.
    - Issue 10: now returning a Generator instead of an array when parsing into
      nodes.
    - Cleaning up the README.rst and comments a bit.

- 2014/11/24: 0.0.7 release.
    - Issue 7: reversing decision in Issue #5: input/output in Python 2
      is a string in the encoding of the mecab lib; while in Python 3,
      both input/output will be in Unicode.
    - Minor refactoring to tighten up attribute visibility in MeCab and
      DictionaryInfo
    - Now testing on Ubuntu 14 (32-bit) as well.

- 2014/11/21: 0.0.6 release.
    - Issue #5: All input/output in Unicode; Python 2.7 users take heed.
    - Issue #7: MeCabEnv now uses Windows Registry to discover lib path.

- 2014/11/08: 0.0.5 release.
    - Issue #4: adding lib to natto.mecab.MeCab.
    - Issue #6: adding __del__ to natto.mecab.MeCab.

- 2014/11/08: 0.0.4 release.
    - Minor cosmetic fixes.

- 2014/11/08: 0.0.3 release.
    - Upgrading to beta status
    - String-Unicode conversion to support Python 2 and 3, keeping to principal
      of least surprise with regards to behavior of strings
    - Supporting Python 3.4
    - Supporting Python 3.3
    - Supporting Python 3.2
    - Supporting Python 2.7
    - Node-formatting for node parsing
    - Automagic defaulting for MeCab library path and charset, but allowing for
      overriding with MECAB_PATH and MECAB_CHARSET
    - Tested on Windows XP (32-bit) and Windows 7 (64-bit)
    - Tested on Mac OS Yosemite (64-bit)
    - Tested on FreeBSD 10 (32-bit)

- 2014/10/26: 0.0.2 release.
    - work in progress

- 2014/10/24: Project inception, initial archiving on Bitbucket.
    - Initial archiving 
    - natto-py is to provide an easy-to-use Python binding with MeCab


natto/lattice.py
# -*- coding: utf-8 -*-
'''Internal-use interface to Lattice.'''
import re
from .support import string_support, unicode_support

class Lattice(object):
    '''Wrapper interface to the MeCab Lattice class, used internally by natto-py.'''

    MECAB_ANY_BOUNDARY = 0
    MECAB_TOKEN_BOUNDARY = 1
    MECAB_INSIDE_TOKEN = 2

    def __init__(self, mecab, tagger, ffi, fn_name, envch):
        self.__mecab = mecab
        self.__tagger = tagger
        self.__ffi = ffi
        self.__fn_name = fn_name
        self.__lattice = self.__mecab.mecab_lattice_new()
        self.__bytes2str, self.__str2bytes = string_support(envch)
        self.__str2unicode, self.__unicode2bytes = unicode_support(envch)
        self.__text = None
        self.__nbest = 1

    def __del__(self):
        if hasattr(self, '_Lattice__mecab') and hasattr(self, '_Lattice__lattice'):
            if self.__lattice != self.__ffi.NULL:
                self.__mecab.mecab_lattice_destroy(self.__lattice)
                del self.__lattice

            if hasattr(self, '_Lattice__mecab'):
                del self.__mecab
            if hasattr(self, '_Lattice__tagger'):
                del self.__tagger
            if hasattr(self, '_Lattice__ffi'):
                del self.__ffi
            if hasattr(self, '_Lattice__fn_name'):
                del self.__fn_name
            if hasattr(self, '_Lattice__bytes2str'):
                del self.__bytes2str
            if hasattr(self, '_Lattice__str2bytes'):
                del self.__str2bytes
            if hasattr(self, '_Lattice__str2unicode'):
                del self.__str2unicode
            if hasattr(self, '_Lattice__unicode2bytes'):
                del self.__unicode2bytes

    def __enter__(self):
        return self

    def __exit__(self, type, value, traceback):
        self.__del__()

    def __split(self, pattern, text):
        '''Split text using the given pattern and return list of tuples of
        token and boolean indicating a pattern match.

        :param pattern: regular expression pattern.
        :type pattern: str
        :param text: text to split.
        :type text: str
        '''
        mark = 0

        patt = self.__str2unicode(pattern)
        text = self.__str2unicode(text)

        for token in re.finditer(patt, text):
            if mark < token.start():
                chunk = self.__unicode2bytes(text[mark:token.start()])
                #chunk = text[mark:token.start()]
                yield (chunk, False)
                mark = token.start()
            chunk = self.__unicode2bytes(text[mark:token.end()])
            #chunk = text[mark:token.end()]
            yield (chunk, True)
            mark = token.end()
        if mark < len(text):
            chunk = self.__unicode2bytes(text[mark:])
            #chunk = text[mark:]
            yield (chunk, False)

    def clear(self):
        self.__mecab.mecab_lattice_clear(self.__lattice)

    def size(self):
        return self.__mecab.mecab_lattice_get_size(self.__lattice)

    def is_available(self):
        return self.__mecab.mecab_lattice_is_available(self.__lattice)

    def set_nbest(self, nbest):
        self.__nbest = nbest

    def get_request_type(self):
        return self.__mecab.mecab_lattice_get_request_type(self.__lattice)

    def set_request_type(self, req_type):
        '''Set request type for Lattice-based parsing.

        :param req_type: request type Lattice-based parsing.
        :type req_type: int
        '''
        self.__mecab.mecab_lattice_set_request_type(self.__lattice, req_type)

    def set_sentence(self, text):
        '''Set sentence text for Lattice-based parsing.

        :param text: target sentence for parsing.
        :type text: str
        '''
        print('in set_sentence..')
        print("text going in: {}".format(text))
        self.__text = text
        btext = self.__str2bytes(text)
        print("bytes going in: {}".format(len(btext)))
        self.__mecab.mecab_lattice_set_sentence(self.__lattice, btext)
        #self.__mecab.mecab_lattice_set_sentence2(self.__lattice, btext, len(btext))

    def set_boundary_constraints(self, morpheme_constraint, any_boundary):
        '''Set the morpheme constraint pattern and preferred boundary marker.

        :param morpheme_constraint: regular expression pattern for morpheme
            constraints.
        :type morpheme_constraint: str
        :param any_boundary: if True, use MECAB_ANY_BOUNDARY for default
            boundary marker; else use MECAB_INSIDE_TOKEN.
        :type any_boundary: bool
        '''
        if any_boundary:
            default_boundary = self.MECAB_ANY_BOUNDARY
        else:
            default_boundary = self.MECAB_INSIDE_TOKEN

        bits = list(self.__split(morpheme_constraint, self.__text))
        for b in bits:
            print(len(b[0]))

        foo = self.__mecab.mecab_lattice_get_sentence(self.__lattice)
        raw = self.__ffi.string(foo)
        #print("bytes in lattice? {}".format(len(raw)))
        #print("text returned from lattice: {}".format( self.__bytes2str(raw) ))
        
        # niwa
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 0, self.MECAB_TOKEN_BOUNDARY)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 1, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 2, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 3, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 4, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 5, self.MECAB_INSIDE_TOKEN)

        # ni
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 6, self.MECAB_TOKEN_BOUNDARY)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 7, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 8, self.MECAB_INSIDE_TOKEN)

        # haniwaniwatori
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice, 9, self.MECAB_TOKEN_BOUNDARY)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,10, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,11, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,12, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,13, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,14, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,15, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,16, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,17, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,18, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,19, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,20, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,21, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,22, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,23, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,24, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,25, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,26, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,27, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,28, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,29, self.MECAB_INSIDE_TOKEN)

        # ga iru .
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,30, self.MECAB_TOKEN_BOUNDARY)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,31, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,32, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,33, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,34, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,35, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,36, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,37, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,38, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,39, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,40, self.MECAB_INSIDE_TOKEN)
        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,41, self.MECAB_INSIDE_TOKEN)

#        pos = 0
#        self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,
#                                                           pos,
#                                                           self.MECAB_TOKEN_BOUNDARY)
#
#        for (token, match) in self.__split(morpheme_constraint, self.__text):
#            pos += 1
#            if match:
#                boundary_constraint = self.MECAB_INSIDE_TOKEN
#            else:
#                boundary_constraint = default_boundary
#
#            for _ in range(1, len(token)):
#                self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,
#                                                                   pos,
#                                                                   boundary_constraint)
#                pos += 1
#            self.__mecab.mecab_lattice_set_boundary_constraint(self.__lattice,
#                                                               pos,
#                                                               self.MECAB_TOKEN_BOUNDARY)

    def parse(self):
        '''Return result of applying Lattice-based node parsing. Parse result
        is obtained either thru get_string or by enumerating the nodes in the
        linked list from bos_node. c.f `mecab_parse_lattice in mecab.h <https://code.google.com/p/mecab/source/browse/trunk/mecab/src/mecab.h>`_.

        :return: Lattice-based parse result code.
        '''
        return self.__mecab.mecab_parse_lattice(self.__tagger, self.__lattice)

    def get_string(self):
        '''Return string result of Lattice-based node parsing. c.f
        `mecab_lattice_tostr and mecab_lattice_nbest_tostr in mecab.h <https://code.google.com/p/mecab/source/browse/trunk/mecab/src/mecab.h>`_

        :return: A single string containing the entire MeCab output.
        '''
        args = [self.__lattice]
        if self.__nbest > 1:
            args.append(self.__nbest)
        res = getattr(self.__mecab, self.__fn_name)(*args)
        raw = self.__ffi.string(res)
        return raw
        #return self.__bytes2str(raw)

    def next(self):
        '''Return pointer to next node in linked list when node parsing. c.f
        `mecab_lattice_next in mecab.h <https://code.google.com/p/mecab/source/browse/trunk/mecab/src/mecab.h>`_.

        :return: pointer to beginning-of-sentence node.
        '''
        return self.__mecab.mecab_lattice_next(self.__lattice)

    def bos_node(self):
        '''Return pointer to beginning-of-sentence node when node parsing. c.f
        `mecab_lattice_get_bos_node in mecab.h <https://code.google.com/p/mecab/source/browse/trunk/mecab/src/mecab.h>`_.

        :return: pointer to beginning-of-sentence node.
        '''
        return self.__mecab.mecab_lattice_get_bos_node(self.__lattice)

    def get_error(self):
        '''Return error string from Lattice. c.f. `mecab_lattice_strerror in
        mecab.h <https://code.google.com/p/mecab/source/browse/trunk/mecab/src/mecab.h>`_.

        :return: error string from Lattice.
        '''
        err = self.__lattice.strerror()
        return self.__bytes2str(self.__ffi.string(err))

'''
Copyright (c) 2015, Brooke M. Fujita.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above
   copyright notice, this list of conditions and the
   following disclaimer.

 * Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the
   following disclaimer in the documentation and/or other
   materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'''
